src/Lecture2.hs

```diff
@@ -154,13 +179,105 @@ You're free to define any helper functions.
 -}
 
 -- some help in the beginning ;)
+
+data Tresure = 
+  Gemstone |
+  Armor |
+  Candy
+
+showTresure :: Tresure -> String 
+showTresure tr = case tr of
+  Gemstone -> "Gemstone"
+  Armor -> "Armor"
+  Candy -> "Candy"
+
+-- A chest contains a non-zero amount of gold and a possible treasure.
+--     When defining the type of a treasure chest, you don't know what
+--     treasures it stores insight, so your chest data type must be able
+--     to contain any possible treasure.
+data Chest a = Chest 
+  {
+      chestGold :: Int,
+      chestTresure :: Maybe a
+  }
+
+showChest :: Chest Tresure -> String 
+showChest ch = case chestTresure ch of
+  Just tr -> "Gold = " ++ show (chestGold ch) ++ " Tresure = " ++ showTresure tr
+  Nothing -> "Gold = " ++ show (chestGold ch)
+  
+
+data Reward = Reward
+  {
+    rewardChest :: Chest Tresure,
+    rewardExperience :: Int
+  }
+
+showReward :: Reward -> String
+showReward r = "Experience = " ++ show (rewardExperience r) ++ " Chest = " ++ showChest (rewardChest r)
+
+type Attack = Int 
+type Health = Int 
+type Endurance = Int
+
 data Knight = Knight
-    { knightHealth    :: Int
-    , knightAttack    :: Int
-    , knightEndurance :: Int
+    { 
+      knightHealth    :: Health,
+      knightAttack    :: Attack,
+      knightEndurance :: Endurance
+    }
+
+-- Experience is calculated based on the dragon type. A dragon can be
+--     either red, black or green.
+data DragonType 
+  = Red
+  | Black
+  | Green
+
+data Dragon = Dragon 
+    {
+      dragonType :: DragonType,
+      dragonHealth :: Health,
+      dragonFirePower :: Attack
     }
 
-dragonFight = error "TODO"
+data FigthResult
+  = KnightDead
+  | KnightWin Reward
+  | KnightRun
+
+showFightResult :: FigthResult -> String 
+showFightResult fightResult = case fightResult of
+  KnightDead -> "Knight is dead"
+  KnightRun -> "Knight run"
+  KnightWin r -> "Knight win " ++ showReward r
+
+
+-- You can try it in ghci console with example:
+  -- ivan = Knight 50 10 10
+  -- dracosha = Dragon Black 50 5
+  -- showFightResult (dragonFight ivan dracosha)
+dragonFight :: Knight -> Dragon -> FigthResult
+dragonFight knigth dragon = fight 1 (dragonHealth dragon) (knightHealth knigth) 
+  where 
+    fight :: Int -> Health -> Attack -> FigthResult
+    fight strikeNo dHealth kHealth
```

Comment:
I see that the third argument has type `Attack` but the variable is actually called `kHealth` 🙂 
If you use `newtype` instead of `type`, you'll be able to rely on GHC to solve such problems for you 👍🏻